py/obj.c

```diff
@@ -174,19 +174,32 @@ mp_int_t mp_obj_hash(mp_obj_t o_in) {
         return mp_obj_tuple_hash(o_in);
     } else if (MP_OBJ_IS_TYPE(o_in, &mp_type_type)) {
         return (mp_int_t)o_in;
-    } else if (MP_OBJ_IS_OBJ(o_in)) {
+    } else if (mp_obj_is_instance_type(mp_obj_get_type(o_in))) {
         // if a valid __hash__ method exists, use it
-        mp_obj_t hash_method[2];
-        mp_load_method_maybe(o_in, MP_QSTR___hash__, hash_method);
-        if (hash_method[0] != MP_OBJ_NULL) {
-            mp_obj_t hash_val = mp_call_method_n_kw(0, 0, hash_method);
+        mp_obj_t method[2];
+        mp_load_method_maybe(o_in, MP_QSTR___hash__, method);
+        if (method[0] != MP_OBJ_NULL) {
+            mp_obj_t hash_val = mp_call_method_n_kw(0, 0, method);
             if (MP_OBJ_IS_INT(hash_val)) {
                 return mp_obj_int_get_truncated(hash_val);
             }
+            assert(0);
```

Comment:
If we get here then it's a TypeError("__hash__ method should return an integer").  So we should raise TypeError instead of crashing.  To save bytes we can just raise the "unhashable type" error by jumping to the code at the end of this function.  This is how this function behaved before this patch.